test/main.tf

```diff
@@ -0,0 +1,71 @@
+provider "aws" {
+  region  = "us-west-2"
+  profile = "thunderdome"
+}
+
+resource "aws_cloudwatch_log_group" "group" {
+  name_prefix = "test-subscription"
+}
+
+module "cloudwatch_kinesis_firehose_tf" {
+  source           = "github.com/observeinc/terraform-aws-kinesis-firehose" # tflint-ignore: terraform_module_pinned_source
+  name             = "test-subscription-tf"
+  observe_customer = "101"
+  observe_token    = "fake:token"
+}
+
+module "cloudwatch_logs_subscription" {
+  source           = "../"
+  name             = "test-subscription-tf"
+  filter_name      = "test-tf"
+  kinesis_firehose = module.cloudwatch_kinesis_firehose_tf
+}
+
+resource "aws_iam_role" "subscription_filter" {
+  name_prefix        = "test-subscription"
+  assume_role_policy = <<-EOF
+    {
+      "Version": "2012-10-17",
+      "Statement": [
+        {
+          "Action": "sts:AssumeRole",
+          "Principal": {
+            "Service": "logs.amazonaws.com"
+          },
+          "Effect": "Allow",
+          "Sid": ""
+        }
+      ]
+    }
+  EOF
+}
+
+resource "aws_iam_role_policy_attachment" "subscription_filter" {
```

Comment:
I'd recommend specifying `managed_policy_arns` on the `aws_iam_role` rather than declaring a separate role-policy attachment:

https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role#example-of-exclusive-managed-policies

`managed_policy_arns` declares an authoritative list of policies, meaning, "these are all the attached policies for this role, add them if missing, remove any not declared here." `aws_iam_role_policy_attachment` is only needed when you want a non-exclusive assignment. Meaning that additional policies could be attached to the role externally, without conflicting.

If CloudFormation were attaching policies, you'd need this resource, but that doesn't appear to be the case.

Reply:
Good point. I'll keep this in mind. Do you know if there's a strong reason to prefer setting the policy rather than attaching policies?

Reply:
Or is it more of a small detail that improves dev efficiency when a codebase gets large

Reply:
In general, the reason to prefer declaring the policies along with the role versus attaching policies is that the former provides some guarantee that the Terraform configuration matches the infrastructure, and can be considered a source of truth. An important function of Terraform is [drift detection](https://learn.hashicorp.com/tutorials/terraform/resource-drift), meaning that if the infrastructure has diverged from what's declared, you can detect and correct that. If someone made a change outside of version control, e.g., using the AWS console or CLI, you want to be able to identify and potentially revert it. This is valuable even if your codebase/module is small.

There are valid reasons to use attachments instead. You might have a Terraform root module that creates a role with no privileges and then exposes the role ARN as an `output`. You could then take that role in multiple other Terraform root modules and attach policies, without conflicts. 

In this case, since you're dealing with a testing module, where you call `terraform apply && terraform destroy` in sequence, there is no practical consequence to attachments versus declaring the policies directly as part of the role. You can't have drift when the infrastructure is ephemeral. So this is just an FYI.

Reply:
thanks, so it seems better to default to this since then we end up with better drift detection